Added PYQ Page (Frontend & Backend) with Cloudinary File Uploads

OmAnand857 opened the pull request:
# 🚀 Added PYQ Page (Frontend & Backend) with Cloudinary File Uploads  

## ✨ Features Implemented  
- **Frontend & Backend for PYQs**  
- **Cloudinary Integration** for uploading **PDF, Word, and Image files**  
- **Firestore Database** for storing file links and metadata  
- **Comments System** with upload time tracking  
- **Upvote Feature** for PYQs  
- **Secure Comment Management** (Only comment owner can delete)  

## 🛠️ Setup Instructions  
1. **Setup Cloudinary**:  
   - Create an account at [Cloudinary](https://cloudinary.com/)  
   - Go to **Settings > Security** and enable **Allow unsigned uploads**  
   - Under **Settings > Upload**, allow **RAW file uploads** (Needed for PDFs, Word files, etc.)  

2. **Update Environment Variables**:  
   - Fill in the updated `.env.example` file with your Cloudinary credentials  
   - Rename it to `.env.local`  

3. **Run the Project**  
   ```bash
   npm install
   npm run dev

OmAnand857 committed:
added pyq page with file uploads

gunjjoshi commented on src/pages/pyqs.tsx:
```diff
@@ -1,177 +1,845 @@
+      <div className="border-t pt-4">
+        <div className="flex items-center justify-between mb-2">
+          <h4 className="text-lg font-semibold">
+            Comments ({pyq.comments.length})
```
```suggestion
            Comments ({pyq.comments?.length})
```

gunjjoshi commented on src/pages/pyqs.tsx:
```diff
@@ -1,177 +1,845 @@
+
+        {expandedComments && (
+          <div className="space-y-3 mb-4 max-h-60 overflow-y-auto scrollbar-thin scrollbar-thumb-gray-300 scrollbar-track-gray-100">
+            {pyq.comments.length === 0 ? (
```
```suggestion
            {pyq.comments?.length === 0 ? (
```

gunjjoshi commented on src/pages/pyqs.tsx:
```diff
@@ -1,177 +1,845 @@
+            {pyq.comments.length === 0 ? (
+              <p className="text-gray-500 text-center py-2">No comments yet</p>
+            ) : (
+              pyq.comments.map((comment) => (
```
```suggestion
              pyq.comments?.map((comment) => (
```

gunjjoshi commented on src/pages/pyqs.tsx:
```diff
@@ -1,177 +1,845 @@
+      const pyq = pyqs.find(p => p.id === pyqId);
+      if (!pyq) return;
+  
+      const commentToDelete = pyq.comments.find(comment => comment.id === commentId);
```
```suggestion
      const commentToDelete = pyq.comments?.find(comment => comment.id === commentId);
```

gunjjoshi commented:
@OmAnand857 Similar to the above suggestions, you'll have to add a `?` to conditionally check if the data is existing or not. If it is empty, we'll run into errors such as the following one:

<img width="964" alt="Screenshot 2025-03-04 at 07 13 11" src="https://github.com/user-attachments/assets/8bf83227-909e-47a7-b509-804853abdec7" />

Can you add those for `pyq.title`, `pyq.course`, etc. as well, wherever applicable?

OmAnand857 commented:
how did you get this error ? The course field is mandatory to fill so ideally we should never land in  undefined situation

OmAnand857 committed:
added null checks

gunjjoshi commented:
@OmAnand857 We're not in an 'ideal' situation. What if there are no PYQs initially? That would itself land in these type of errors. It would be better if you can iterate through the changes once again, so that edge cases like this one are covered.

OmAnand857 commented:
@gunjjoshi I have fixed them in latest commit. I have added the optional chaining for null checks as suggested. I was just curious about how to recreate the error cause I couldn't on my system.

gunjjoshi commented on src/components/Common/Navbar.tsx:
```diff
@@ -6,6 +6,7 @@ import { FiLogOut } from 'react-icons/fi'
 import { useRouter } from 'next/router'
 import { useAuth } from '../../contexts/auth'
 import Image from 'next/image'
+import { clear } from 'console'
```
```suggestion
```

gunjjoshi replied:
This isn't being used anywhere, can be removed.

gunjjoshi committed:
Update src/components/Common/Navbar.tsx

gunjjoshi commented:
LGTM. Thanks @OmAnand857!
